### DISCUSSION TITLE
Is Netfox a good choice for physics based 3D environments?

### DISCUSSION THREAD

**Discussion by blackears** [foxssake/netfox#272] [General]
I'm interested in using Netfox for a physically simulated 3D environment where several players would be playing a sports game involving pushing a RigidBody ball around a playing field.  From what I've read in the docs and on other forums, I'm getting the impression Netfox isn't good at this because Godot's current physics engine implementation doesn't allow for deterministic behavior.  On the other hand, Netfox does come with that example where two players are jumping around a 3D landscape and it seems to work.

I'm wondering if it makes sense to try and create a 3D multiplayer physics based ball game using Netfox, or if this is something that is outside of Netfox/Godot's capabilities.

**Comment by albertok** [comment]
@blackears 

The short of it is that as you suspect - Godot doesn't make it easy but it's not impossible with compromises.

Godot's physics aren't deterministic and there's no interface to re-simulate physics multiple times in a render frame. Not many game engine's can for that matter.

So RigidBodies can't really participate in rollback code but it is possible to do state syncing so that the physics simulations are somewhat in sync across clients.

Netfox doesn't have a feature for this yet but it has been talked about - see https://github.com/foxssake/netfox/issues/265 . You basically copy the physics states from the server and apply them to the clients.

The example Netfox comes with doesn't use RigidBodies. It uses more predictable nodes like CharacterBody3D and ShapeCast3D which can be forced to repeat movement and collision checks several times per frame.

**Reply by blackears** [reply]
I just tried the script you posted on #265 and it seems to work really well.  Thanks!  Some of the comments there talk about making it work with the TickInterpolator.  I was wondering if this script is in a 'final' stage or if there is some more work you're thinking of doing with it?

**Comment by TheYellowArchitect** [comment]
I want to add that you can have deterministic physics if physics are server-based, regardless of what engine or physics you use.
Though, I have to mention Jolt (it will be the official Godot physics in the next godot versions)
Apply it atop of Netfox - assuming you apply the same input on the same starting state - it should be deterministic. Jolt itself is deterministic. Jolt plugin says "not sure" because no one has disproven the godot implementation is not deterministic.

>Godot's physics aren't deterministic and there's no interface to re-simulate physics multiple times in a render frame. Not many game engine's can for that matter.

See https://github.com/godotengine/godot/pull/76462

**Reply by blackears** [reply]
I don't know if Jolt is deterministic or not, but in my simulations were I don't try to sync my RigidBody and just let the physics engine do it's thing, the client and server quickly diverge.  Possibly due to slight differences in the position and velocity of the player bodies.

**Reply by albertok** [reply]
@TheYellowArchitect 

Jolt the physics engine itself is deterministic but GodotJolt makes it clear it can't guarantee it. https://github.com/godot-jolt/godot-jolt?tab=readme-ov-file#what-about-determinism

After the issues we solved with the time sync I can see why. Your frame rendering rate directly impacts simulations, if someone frame skips for whatever reason the simulations will be off.

**Comment by blackears** [comment]
Okay.  I've put together a simple 3D networked game using Netfox and Jolt.  I don't mind if the physics are all done on the server and broadcast to the clients, but I'm having trouble getting that to work.

The players are CharacterBody3D and can move around on both the server and client smoothly.  However, I've also added a RigidBody ball I want them to push around and right now, colliding with the ball does not cause it to move.  I've added a RollbackSynchronizer and TickInterpolator to the rigid body and have them serving the transform, linear_velocity and angular_velocity properties.  I would have hoped that this would simulate the physics on the host and then the client would try to sync to that, but instead the player bodies just collide with the RigidBody as if it were a StaticBody.  If I delete the RollbackSynchronizer and TickInterpolator, the physics works again, but the host and client are completely out of sync.

I'm using this code at the end of the player update to push the ball:

```
       ...

	velocity *= NetworkTime.physics_factor
	move_and_slide()
	velocity /= NetworkTime.physics_factor

	for i in get_slide_collision_count():
		var c = get_slide_collision(i)
		if c.get_collider() is RigidBody3D:
			c.get_collider().apply_central_impulse(-c.get_normal() * push_force)
```

**Reply by albertok** [reply]
@blackears 

It's a really hard problem, please keep us up to date with how you go.

I'll be tackling similar problems soon. Have you by any chance tried making the clients have a regular Node instead of a RigidBody? 

The alternative is to just fake physics - use CharacterBody3D on server and clients and apply velocities to them from collisions.

https://x.com/StayAtHomeDev/status/1824178912102175045

**Reply by blackears** [reply]
No, I haven't.  Although set_state() in your script doesn't seem to be being called, so I'm wondering if it might be the MultiplayerSynchronizer alone which is doing it.  I ended up adding the position, rotation, linear_velocity and angular_velocity to the MultiplayerSynchronizer before it started working.  Although when I look at your source, I think that it is trying to do something similar.  Is there something you need to do to activate it?

**Comment by ServerusSnap** [comment]
https://github.com/foxssake/netfox/releases/tag/v1.35.3
